### PR TITLE
std.typecons: Make RefCounted pure nothrow @nogc when possible

### PR DESCRIPTION
**Pull request by nordlow**
This makes `RefCounted!T` `pure/nothrow/@nogc` when possible.

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -4970,6 +4970,15 @@ enum RefCountedAutoInitialize
     yes,
 }
 
+// TODO can we make these imports local to RefCounted?
+extern(C)
+{
+    // qualified C memory allocations
+    private @safe pure nothrow @nogc:
+    void* malloc(size_t size);
+    void free(void* ptr);
```
Comment (andralex) [inline std/typecons.d]: free is not and cannot be safe

Comment (nordlow) [inline std/typecons.d]: Ok. But its usage here can be `@trusted`.

Comment (andralex) [inline std/typecons.d]: I see, you made it private too. Hmmm...

**Comment by andralex** [comment]
OK how about you keep the safe but for now drop the pure. The notion of marking `malloc` as pure is disquieting - it never returns the same `void*` for the same number. Then we'll see how we can do about pure. Thanks!

**Comment by jmdavis** [comment]
> The notion of marking malloc as pure is disquieting - it never returns the same void* for the same number.

I would point out that we're in the same boat with the GC and `new`. It's `pure` and doesn't necessarily return the same value - just an equivalent one. The only real difference is that with the GC, we have something to manage freeing that memory, whereas since you have to worry about it with `malloc`, you have to be concerned about whether you really do get the same value so that you don't free it twice.

So, I don't know what the right answer is, but there's definitely an argument that `malloc` should be considered `pure` given what we do with `new`.

**Comment by andralex** [comment]
@jmdavis new is special so we can special case it

**Comment by nordlow** [comment]
AFAICT, C-style memory allocations cannot be qualified correctly in D using `pure`. For instance if we declare `malloc` as `pure` the optimizer will optimize away successive calls to `malloc` which is incorrect. The return value of `malloc` shall not in the `RefCounted` case have any side-effect on the program only the values that are stored in it. This semantics cannot currently be expressed.

Instead we have to qualify the code that it's _using_ them and privately import `*alloc` and `free` locally for each use. I believe the same holds for `GC.addRange`.

**Comment by aG0aep6G** [comment]
@andralex:

> The notion of marking malloc as pure is disquieting - it never returns the same void* for the same number.

That's not a requirement of D's `pure` attribute. Or if it is, we have to fix the compiler to enforce it and probably break lots of code.

> new is special so we can special case it

GC.malloc and GC.free aren't really special, though. They can be used just like C's malloc and free. Yet GC.malloc and GC.free are `pure` and the C functions are not. I don't see the reason for the disparity.

GC.malloc being (weakly) `pure` is ok because void* has a mutable indirection, so the compiler must not reuse the result (but: [issue 15862](https://issues.dlang.org/show_bug.cgi?id=15862)). GC.free being (weakly) `pure` is ok because the visible effect it has is on its mutable argument. The same points apply to C's malloc and free.

See also <http://forum.dlang.org/post/hrwnhprgcjlbypaecpzg@forum.dlang.org>.

@nordlow:

> if we declare malloc as pure the optimizer will optimize away successive calls to malloc

Actually, the compiler must not do that. It does right now, but that's [issue 15862](https://issues.dlang.org/show_bug.cgi?id=15862).

**Comment by andralex** [comment]
@aG0aep6G I think it's a slippery slope that GC.malloc and free are marked as pure. I don't like that, and I don't like that used as a precedent.

There must be some acknowledgment at a point that our type system has limitations that doesn't allow it to express certain valid programs.

**Comment by aG0aep6G** [comment]
> @aG0aep6G I think it's a slippery slope that GC.malloc and free are marked as pure. I don't like that, and I don't like that used as a precedent.

This needs to be sorted out. Either it's wrong that GC.malloc and GC.free are marked `pure`, or we should be able to mark C's malloc and free too. Either we have a disaster waiting to happen with the GC functions, or we're needlessly preventing code from being `pure`.

> There must be some acknowledgment at a point that our type system has limitations that doesn't allow it to express certain valid programs.

Ok, but I don't think it's a good idea to draw that line between GC.malloc and C's malloc.

**Comment by andralex** [comment]
Hmmm, this notion of a function that returns data with mutable indirection is very interesting. In particular, if the data returned has mutable indirections leading to values that cannot be present in the parameters, that means that's "fresh" data, i.e. newly allocated. Consider:

```d
pure int[] fun(float[]);
```

Without even looking at the definition, the compiler can infer that the array of integers is freshly-allocated for the simple reason there's no other place it could come from :o). Being able to characterize "fresh" without making language addition can be a big win.

For now, I agree that fixing https://issues.dlang.org/show_bug.cgi?id=15862 would allow us to make malloc pure.

**Comment by aG0aep6G** [comment]
> In particular, if the data returned has mutable indirections leading to values that cannot be present in the parameters, that means that's "fresh" data, i.e. newly allocated.

Yeah, dmd uses this to allow implicit conversion to immutable when the value comes from a pure function (and other requirements are met). This example compiles with `pure` and is correctly rejected without it:

```d
int* f() pure { return new int; }
void main() { immutable p = f(); }
```

> Consider:
> 
>     pure int[] fun(float[]);
> 
> Without even looking at the definition, the compiler can infer that the array of integers is freshly-allocated for the simple reason there's no other place it could come from :o).

I'm not so sure about that example. fun is allowed to cast the argument to int[], isn't it? Then again, dmd seems to disregard that possibility, too. This is currently accepted:

```d
pure int[] fun(float[] arg) { return cast(int[]) arg; }
void main() { float[] a = [1, 2, 3]; immutable i = fun(a); }
```

If dmd is correct here, that would mean casting from one array type to another is invalid (aka has undefined behavior), wouldn't it? That doesn't seem right. I'd say this is another bug in dmd, and it cannot be assumed that fun returns a unique result. Filed as [issue 16585](https://issues.dlang.org/show_bug.cgi?id=16585).

**Comment by nordlow** [comment]
@andralex I thought you had decided upon non-safety in https://github.com/dlang/phobos/pull/4692

**Comment by andralex** [comment]
>@andralex I thought you had decided upon non-safety in #4692 ...

@nordlow not sure I understand?

**Comment by wilzbach** [comment]
> For now, I agree that fixing https://issues.dlang.org/show_bug.cgi?id=15862 would allow us to make malloc pure.

The bug has been fixed ;-)

https://github.com/dlang/dmd/pull/6197

So what's still blocking this?

**Comment by nordlow** [comment]
The `RefCounted` destructor calls `free`. It too has to be marked as `pure`. I can look into making it pure aswell and then go forward with this.

**Comment by schveiguy** [comment]
I think instead of marking `free` pure to make this PR successful, we can work around the purity.

Make a private binding to `free`, something like:

```D
pragma(mangle, "free") extern(C) private purefree(void *mem) pure;
```

Then you can call it in one place where we can control the context, and not expose the whole world.

**Comment by nordlow** [comment]
@schveiguy, It doesn't. I just thought making core.stdc pure was the direction @andralex wanted. Andrei?

**Comment by andralex** [comment]
Let's go with the private declaration for now. Yes, in the future we do want to make functions like `free` pure.

**Comment by nordlow** [comment]
Ok, does that include `gc_addRange` and `gc_removeRange`, @andralex ?

**Comment by andralex** [comment]
1. yes those should be fine too (or at least subject to the same subtleties)
2. @schveiguy 's seems a bit clearer

**Comment by nordlow** [comment]
Made it so, @andralex

File: std/typecons.d [hunk 2]
```diff
@@ -4986,31 +4994,27 @@ if (!is(T == class) && !(is(T == interface)))
...
-            import core.memory : GC;
-            import core.stdc.stdlib : malloc;
             import std.conv : emplace;
 
             _store = cast(Impl*)malloc(Impl.sizeof);
             if (_store is null)
                 onOutOfMemoryError();
             static if (hasIndirections!T)
-                GC.addRange(&_store._payload, T.sizeof);
+                pure_gc_addRange(&_store._payload, T.sizeof);
```
Comment (andralex) [inline std/typecons.d]: line not covered

File: std/typecons.d [hunk 3]
```diff
@@ -4972,9 +4972,17 @@ struct RefCounted(T, RefCountedAutoInitialize autoInit =
         RefCountedAutoInitialize.yes)
 if (!is(T == class) && !(is(T == interface)))
 {
+    extern(C) private pure nothrow @nogc static // TODO remove pure when https://issues.dlang.org/show_bug.cgi?id=15862 has been fixed
+    {
+        pragma(mangle, "free") void pure_free( void *ptr );
```
Comment (andralex) [inline std/typecons.d]: unusual naming convention (`pureFree` would be preferred)

File: std/typecons.d [hunk 4]
```diff
@@ -4972,9 +4972,17 @@ struct RefCounted(T, RefCountedAutoInitialize autoInit =
...
+        pragma(mangle, "gc_addRange") void pure_gc_addRange( in void* p, size_t sz, const TypeInfo ti = null );
```
Comment (andralex) [inline std/typecons.d]: Eh, apparently not that unusual :)

**Comment by andralex** [comment]
Don't worry about the name, let's get the coverage. Thx!

File: std/typecons.d (hunk 2 above)
Comment (nordlow) [inline std/typecons.d]: Done.

File: std/typecons.d [hunk 5]
```diff
@@ -5307,6 +5310,8 @@ assert(refCountedStore.isInitialized)).
     RefCounted!int b;
     b = a; //This should not assert either
     assert(b == 5);
+
+    RefCounted!(int*) c;
```
Comment (nordlow) [inline std/typecons.d]: This makes line 5003 covered.

**Comment by nordlow** [comment]
Coverage fixed.

File: std/typecons.d [hunk 6]
```diff
@@ -5298,7 +5301,15 @@ assert(refCountedStore.isInitialized)).
     assert(val == 4);
 }
 
-@system unittest
+// gc_addRange coverage
+@system pure unittest
+{
+    struct S { int* p; }
+
+    auto s = RefCounted!S(null);
```
Comment (nordlow) [inline std/typecons.d]: This makes line 5003 covered.

**Comment by nordlow** [comment]
Next up is `Mallocator`. Right, @andralex?

File: std/typecons.d (hunk 4 above)
Comment (schveiguy) [inline std/typecons.d]: Kill the TODO, that bug has been fixed, and you still need `pure` anyway.

**Comment by schveiguy** [comment]
I think this looks good. Let's wait and see whether dlang/druntime#1735 is merged or not. If so, then you need to have a `pure_malloc` as well.

**Comment by nordlow** [comment]
If https://github.com/dlang/druntime/pull/1735 gets merged we can use `pureMalloc` instead, right?

**Comment by andralex** [comment]
@nordlow I guess this is next!

**Comment by nordlow** [comment]
So we should make use of `pureMalloc` and `pureRealloc`. What about purity of `gc_addRange` and `gc_removeRange`? Should we use locally purefied versions of them?

**Comment by andralex** [comment]
local is fine for now

**Comment by nordlow** [comment]
Ok, in that case this is ready, @andralex !

**Comment by andralex** [comment]
@nordlow you have style failures, see ci/circleci

**Comment by JackStouffer** [comment]
@andralex That's actually a dscanner error. It isn't working correctly with the new DIP1000 syntax.

**Comment by nordlow** [comment]
So what to do, @andralex ?

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by nordlow** [comment]
Great!

**Pull request merged**